File: scratch/check_syntax.py
import re
import subprocess
import os

def check():
    html_path = r"c:\Users\Nam\OneDrive\Desktop\vinastudy-bot\tower_defense.html"
    with open(html_path, "r", encoding="utf-8") as f:
        content = f.read()

    # Find the main <script> tag content (after the SoundFX or near the end)
    # Let's extract all content inside <script>...</script>
    scripts = re.findall(r"<script>(.*?)</script>", content, re.DOTALL)
    if not scripts:
        print("No script tags found!")
        return

    # Write each script block to a temporary file and check it
    os.makedirs("scratch", exist_ok=True)
    for idx, script_content in enumerate(scripts):
        temp_js = f"scratch/temp_check_{idx}.js"
        with open(temp_js, "w", encoding="utf-8") as f_out:
            f_out.write(script_content)

        print(f"Checking script block {idx}...")
        result = subprocess.run(["node", "--check", temp_js], capture_output=True, text=True)
        if result.returncode != 0:
            print(f"❌ Syntax Error found in script block {idx}!")
            print(result.stderr)
            # Map temp JS line numbers back to original HTML line numbers
            # Find the start line of the script tag in the HTML
            lines = content.splitlines()
            script_start_line = 0
            # A simple search for the idx-th <script> tag
            script_count = 0
            for line_idx, line in enumerate(lines):
                if "<script>" in line:
                    if script_count == idx:
                        script_start_line = line_idx + 1
                        break
                    script_count += 1
            
            print(f"Original <script> tag starts at line: {script_start_line}")
            # Parse line number from error
            err_line_match = re.search(r"temp_check_\d+\.js:(\d+)", result.stderr)
            if err_line_match:
                js_err_line = int(err_line_match.group(1))
                html_err_line = script_start_line + js_err_line
                print(f"Approximate HTML line of error: {html_err_line}")
        else:
            print(f"✅ Script block {idx} is syntactically correct!")

if __name__ == "__main__":
    check()

File: scratch/check_syntax.js
const fs = require('fs');
const path = require('path');
const vm = require('vm');

function check() {
    const htmlPath = path.join(__dirname, '..', 'tower_defense.html');
    const content = fs.readFileSync(htmlPath, 'utf8');

    // Simple script tag regex extractor
    const scriptRegex = /<script>([\s\S]*?)<\/script>/gi;
    let match;
    let idx = 0;

    while ((match = scriptRegex.exec(content)) !== null) {
        const jsCode = match[1];
        
        // Find the line number in HTML where this script starts
        const htmlBefore = content.substring(0, match.index);
        const startLineNum = htmlBefore.split('\n').length;

        console.log(`Checking script block ${idx} (starts at HTML line ${startLineNum})...`);
        
        try {
            // Attempt to compile the script content using node's vm module
            new vm.Script(jsCode, { filename: `tower_defense.html_script_${idx}.js` });
            console.log(`✅ Script block ${idx} is syntactically correct!`);
        } catch (err) {
            console.log(`❌ Syntax Error found in script block ${idx}!`);
            console.log(err.stack || err);
            
            // Extract the line number in the script and calculate HTML line number
            if (err.stack) {
                const lineMatch = err.stack.match(/tower_defense\.html_script_\d+\.js:(\d+)/);
                if (lineMatch) {
                    const jsErrLine = parseInt(lineMatch[1]);
                    const htmlErrLine = startLineNum + jsErrLine - 1;
                    console.log(`\n👉 Exact HTML line of syntax error is approximately: ${htmlErrLine}\n`);
                    
                    // Show a few lines of code around the error
                    const htmlLines = content.split('\n');
                    const startShow = Math.max(0, htmlErrLine - 5);
                    const endShow = Math.min(htmlLines.length - 1, htmlErrLine + 5);
                    console.log("Code snippet around error:");
                    for (let l = startShow; l <= endShow; l++) {
                        const marker = (l === htmlErrLine - 1) ? "👉 " : "   ";
                        console.log(`${marker}${l + 1}: ${htmlLines[l]}`);
                    }
                }
            }
        }
        idx++;
    }
}

check();
